mpt/impl/smtp/cassandra-rabbitmq-object-storage/src/test/java/org/apache/james/mpt/smtp/CassandraRabbitMQAwsS3SmtpTestRule.java

```diff
@@ -0,0 +1,81 @@
+ * distributed with this work for additional information        *
+ * regarding copyright ownership.  The ASF licenses this file   *
+ * to you under the Apache License, Version 2.0 (the            *
+ * "License"); you may not use this file except in compliance   *
+ * with the License.  You may obtain a copy of the License at   *
+ *                                                              *
+ *   http://www.apache.org/licenses/LICENSE-2.0                 *
+ *                                                              *
+ * Unless required by applicable law or agreed to in writing,   *
+ * software distributed under the License is distributed on an  *
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY       *
+ * KIND, either express or implied.  See the License for the    *
+ * specific language governing permissions and limitations      *
+ * under the License.                                           *
+ ****************************************************************/
+package org.apache.james.mpt.smtp;
+
+import java.util.function.BiFunction;
+
+import org.apache.commons.configuration.DefaultConfigurationBuilder;
+import org.apache.james.CassandraJamesServerMain;
+import org.apache.james.CleanupTasksPerformer;
+import org.apache.james.GuiceJamesServer;
+import org.apache.james.backend.rabbitmq.DockerRabbitMQSingleton;
+import org.apache.james.backends.cassandra.init.configuration.ClusterConfiguration;
+import org.apache.james.dnsservice.api.DNSService;
+import org.apache.james.dnsservice.api.InMemoryDNSService;
+import org.apache.james.modules.TestAwsS3BlobStoreModule;
+import org.apache.james.modules.TestRabbitMQModule;
+import org.apache.james.modules.blobstore.BlobStoreChoosingModule;
+import org.apache.james.modules.protocols.SmtpGuiceProbe.SmtpServerConnectedType;
+import org.apache.james.modules.rabbitmq.RabbitMQModule;
+import org.apache.james.modules.server.CamelMailetContainerModule;
+import org.apache.james.queue.api.MailQueueItemDecoratorFactory;
+import org.apache.james.queue.api.RawMailQueueItemDecoratorFactory;
+import org.apache.james.server.core.configuration.Configuration;
+import org.apache.james.util.Host;
+import org.junit.rules.TemporaryFolder;
+
+import com.github.fge.lambdas.Throwing;
+
+public abstract class CassandraRabbitMQAwsS3SmtpTestRule {
+    public static SmtpTestRule create(SmtpServerConnectedType smtpServerConnectedType, Host cassandraHost) {
+        BiFunction<TemporaryFolder, InMemoryDNSService, GuiceJamesServer> createJamesServer = Throwing
+            .<TemporaryFolder, InMemoryDNSService, GuiceJamesServer>biFunction((folder, dnsService) -> createJamesServer(cassandraHost, folder, dnsService))
+            .sneakyThrow();
+
+        return new SmtpTestRule(smtpServerConnectedType, createJamesServer);
+    }
+
+    private static GuiceJamesServer createJamesServer(Host cassandraHost, TemporaryFolder folder, DNSService dnsService) throws Exception {
```

Comment:
I still don't agree with the fact we define rules for things setup instead of lifecycle management.
This server description doesn't belong to a Rule.

Reply:
> I still don't agree with the fact we define rules for things setup instead of lifecycle management.

Do you have an example in tests please?

> This server description doesn't belong to a Rule.

It does not, it parametrizes it.

Reply:
I've been confused by the names. This class is not a Rule but a RuleFactory, right?